src/PIL/DdsImagePlugin.py

```diff
@@ -237,6 +236,40 @@ def load_seek(self, pos):
         pass
 
 
+class DdsRgbDecoder(ImageFile.PyDecoder):
+    _pulls_fd = True
+
+    def decode(self, buffer):
+        bitcount, masks = self.args
+
+        # Some masks will be padded with zeros, e.g. R 0b11 G 0b1100
+        # Calculate how many zeros each mask is padded with
+        mask_offsets = []
+        # And the maximum value of each channel without the padding
+        mask_totals = []
+        for mask in masks:
+            offset = 0
+            if mask != 0:
+                while mask >> (offset + 1) << (offset + 1) == mask:
+                    offset += 1
+            mask_offsets.append(offset)
+            mask_totals.append(mask >> offset)
+
+        data = bytearray()
+        bytecount = bitcount // 8
+        while len(data) < self.state.xsize * self.state.ysize * len(masks):
```

Comment:
I personally think the current version is simpler to read, and expect the performance difference to be negligible.

The current version is also used in a few other places in Pillow - but at the same time, that's probably because I wrote them.

https://github.com/python-pillow/Pillow/blob/316f39702c16ed97a95b971cb4aad8a87a26f9df/src/PIL/QoiImagePlugin.py#L55